core/src/main/scala/org/apache/spark/util/ListenerBus.scala

```diff
@@ -17,53 +17,222 @@
+          task.run()
+          processingEvent.set(false)
+      }
+      })
+    } catch {
+      case e: RejectedExecutionException =>
+        droppedEventsCounter.incrementAndGet()
+        if (System.currentTimeMillis() - lastReportTimestamp >= 60 * 1000) {
+          val droppedEvents = droppedEventsCounter.get
+          // There may be multiple threads trying to decrease droppedEventsCounter.
+          // Use "compareAndSet" to make sure only one thread can win.
+          // And if another thread is increasing droppedEventsCounter, "compareAndSet" will fail and
+          // then that thread will update it.
+          if (droppedEventsCounter.compareAndSet(droppedEvents, 0)) {
+            droppedEventsCounter.set(0)
+            val prevLastReportTimestamp = lastReportTimestamp
+            lastReportTimestamp = System.currentTimeMillis()
+            logError(s"Dropping $droppedEvents SparkListenerEvent since " +
+              new java.util.Date(prevLastReportTimestamp) +
+              " because no remaining room in event queue. This likely means" +
+              s" $listenerName event processor is too slow and cannot keep up " +
+              "with the rate at which tasks are being started by the scheduler.")
+          }
+        }
+    }
+  }
+
+  private[this] def isProcessingEvent: Boolean = processingEvent.get()
+
+  def isEmpty: Boolean = {
+    executorService.getQueue.size() == 0 && !isProcessingEvent
+  }
+
+  def stop(): Unit = {
+    executorService.shutdown()
+    executorService.awaitTermination(10, TimeUnit.SECONDS)
+  }
+}
 
 /**
  * An event bus which posts events to its listeners.
  */
 private[spark] trait ListenerBus[L <: AnyRef, E] extends Logging {
 
-  // Marked `private[spark]` for access in tests.
-  private[spark] val listeners = new CopyOnWriteArrayList[L]
+  // Cap the capacity of the event queue so we get an explicit error (rather than
+  // an OOM exception) if it's perpetually being added to more quickly than it's being drained.
+  protected def eventQueueSize = 10000
+  private val eventGroupToEventExecutors =
+    new ConcurrentHashMap[String, ListenerEventExecutor[L]] ()
+
+  // Indicate if `start()` is called
+  private val started = new AtomicBoolean(false)
+  // Indicate if `stop()` is called
+  private val stopped = new AtomicBoolean(false)
 
   /**
    * Add a listener to listen events. This method is thread-safe and can be called in any thread.
    */
-  final def addListener(listener: L): Unit = {
-    listeners.add(listener)
+  final def addListener(
+    listener: L, eventListenerGroup: String = ListenerEventExecutor.DefaultEventListenerGroup):
+  Unit = synchronized {
+    var listenerEventExecutor = eventGroupToEventExecutors.get(eventListenerGroup)
+    if (listenerEventExecutor == null) {
+      listenerEventExecutor =
+        new ListenerEventExecutor[L](listener.getClass.getName, eventQueueSize)
+      eventGroupToEventExecutors.put(eventListenerGroup, listenerEventExecutor)
```

Comment:
I think that if you make `start` also synchronized, and move this after the `started.get()` check below, you can avoid all the `waitForStart()` stuff in the dispatcher class. I haven't fully checked if that will work, but if it does, it will greatly simplify the code in the dispatcher class (no atomic booleans, locks, signals, etc).